model/src/signature.rs

```diff
@@ -0,0 +1,208 @@
+    signing_scheme: EcdsaSigningScheme,
+    domain_separator: &DomainSeparator,
+    struct_hash: &[u8; 32],
+) -> [u8; 32] {
+    match signing_scheme {
+        EcdsaSigningScheme::Eip712 => hashed_eip712_message(domain_separator, struct_hash),
+        EcdsaSigningScheme::EthSign => hashed_ethsign_message(domain_separator, struct_hash),
+    }
+}
+
+impl EcdsaSignature {
+    /// r + s + v
+    pub fn to_bytes(self) -> [u8; 65] {
+        let mut bytes = [0u8; 65];
+        bytes[..32].copy_from_slice(self.r.as_bytes());
+        bytes[32..64].copy_from_slice(self.s.as_bytes());
+        bytes[64] = self.v;
+        bytes
+    }
+
+    pub fn from_bytes(bytes: &[u8; 65]) -> Self {
+        EcdsaSignature {
+            r: H256::from_slice(&bytes[..32]),
+            s: H256::from_slice(&bytes[32..64]),
+            v: bytes[64],
+        }
+    }
+
+    pub fn validate(
+        &self,
+        signing_scheme: EcdsaSigningScheme,
+        domain_separator: &DomainSeparator,
+        struct_hash: &[u8; 32],
+    ) -> Option<H160> {
+        let message = hashed_signing_message(signing_scheme, domain_separator, struct_hash);
+        let recovery = Recovery::new(message, self.v as u64, self.r, self.s);
+        let (signature, recovery_id) = recovery.as_signature()?;
+        signing::recover(&message, &signature, recovery_id).ok()
+    }
+
+    pub fn sign(
+        signing_scheme: EcdsaSigningScheme,
+        domain_separator: &DomainSeparator,
+        struct_hash: &[u8; 32],
+        key: SecretKeyRef,
+    ) -> Self {
+        let message = hashed_signing_message(signing_scheme, domain_separator, struct_hash);
+        // Unwrap because the only error is for invalid messages which we don't create.
+        let signature = key.sign(&message, None).unwrap();
+        Self {
+            v: signature.v as u8,
+            r: signature.r,
+            s: signature.s,
+        }
+    }
+}
+
+impl Serialize for EcdsaSignature {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut bytes = [0u8; 2 + 65 * 2];
+        bytes[..2].copy_from_slice(b"0x");
+        // Can only fail if the buffer size does not match but we know it is correct.
+        hex::encode_to_slice(&self.to_bytes(), &mut bytes[2..]).unwrap();
+        // Hex encoding is always valid utf8.
+        let str = std::str::from_utf8(&bytes).unwrap();
+        serializer.serialize_str(str)
+    }
+}
+
+impl<'de> Deserialize<'de> for EcdsaSignature {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct Visitor {}
+        impl<'de> de::Visitor<'de> for Visitor {
+            type Value = EcdsaSignature;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                write!(
+                    formatter,
+                    "the 65 ecdsa signature bytes as a hex encoded string"
```

Comment:
ubernit: maybe include the expected order of the bytes here (`r .. s .. v`) and that V is in Electrum notation (i.e. 27 or 28 and not 0 or 1).